### PR TITLE
Add API to construct the functional variant of an op

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #102293
* #102292

`register_functional_op`:
- constructs the functional variant of an op
- registers a functionalization kernel to the op

To get this to work:
- `register_functional_op` makes assumptions that it checks about the
op's schema. In particular, the op is not allowed to return anything it
mutates. We can relax these constraints in the future.
- We add a "boxed" python functionalization kernel that handles this
case.

I'm not actually sure (or convinced) this should be public API or how
it should work. If we want this to be public, then it should probably be
a torch.library API, but does that also mean we should give the same
lifetime guarantees? If so, then it would be up to the user to construct
a Library object to actually register the functional variant onto.

Test Plan:
- new tests

### CODE REVIEW

**Comment by ezyang** [comment]
I'm deferring this review to @bdhirsh

File: torch/_custom_op/functional.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        if ret.annotation is not None:
+            raise NotImplementedError(
+                "NYI: register_functional_op(op) where op returns a mutated or aliased value. "
+                "Please file an issue (and as a workaround, modify your operator to "
+                "not return the mutated value or aliases)")
+    for arg in schema.arguments.flat_all:
+        # construct_functionalization_kernel assumes this for simplicity
+        if arg.type.is_tensor_like() and arg.type != BaseType(BaseTy.Tensor):
+            raise NotImplementedError(
+                "NYI: register_functional_op(op) where op accepts Optional or List of tensors."
```
Comment (bdhirsh) [inline torch/_custom_op/functional.py:160]: nit: in theory it's probably fine if the tensorlist / optional tensor arg doesn't have mutability annotations (but if you'd rather wait for someone to file before handling all of this np).

File: test/test_python_dispatch.py [hunk 2]
```diff
@@ -41,9 +43,13 @@ def test_call_boxed(self) -> None:
 class TestPythonRegistration(TestCase):
     test_ns = '_test_python_registration'
 
+    def setUp(self):
+        self.keep_alive_len = len(torch._custom_op.functional.keep_alive)
+
     def tearDown(self):
         if hasattr(torch.ops, self.test_ns):
             del torch.ops._test_python_registration
+        del torch._custom_op.functional.keep_alive[self.keep_alive_len:]
```
Comment (bdhirsh) [inline test/test_python_dispatch.py:52]: nice

**Review by bdhirsh — APPROVED** [review]
really cool!

**Pull request closed without merging**